subprojects/model-core/src/integTest/groovy/org/gradle/api/provider/PropertyAssignmentIntegrationTest.groovy

```diff
@@ -0,0 +1,405 @@
+        "File = File"                                   | "File"         | "DirectoryProperty" | 'file("$buildDir/out")'                  | "Result: out"                        | "Result: out"
+        "File = Provider<File>"                         | "File"         | "DirectoryProperty" | 'provider { file("$buildDir/out") }'     | failsWithCause("Cannot cast object") | failsWithCause("Cannot get the value of task ':myTask' property 'input'")
+        "File = Object"                                 | "File"         | "DirectoryProperty" | 'new MyObject("out")'                    | failsWithCause("Cannot cast object") | failsWithCause("Cannot set the value of task ':myTask' property 'input'")
+    }
+
+    def "test Kotlin object types assignment for #description"() {
+        given:
+        file("gradle.properties") << "\nsystemProp.org.gradle.unsafe.kotlin.assignment=true\n"
+        def buildFileDefinition = """
+            enum class MyEnum {
+                YES, NO
+            }
+
+            class MyObject(val value: String) {
+                override fun toString() = value
+            }
+
+            abstract class MyTask : DefaultTask() {
+                @get:Internal
+                {inputDeclaration}
+
+                @TaskAction
+                fun run() {
+                    when (val anyInput = input as Any?) {
+                        is DirectoryProperty -> println("Result: " + anyInput.get().asFile.name)
+                        is File -> println("Result: " + anyInput.name)
+                        is Property<*> -> println("Result: " + anyInput.get())
+                        else -> println("Result: " + anyInput)
+                    }
+                }
+            }
+
+            tasks.register<MyTask>("myTask") {
+                input = {inputValue}
+            }
+        """
+
+        when:
+        buildKotlinFile.text = buildFileDefinition
+            .replace("{inputDeclaration}", "var input: $eagerInputType? = null")
+            .replace("{inputValue}", inputValue)
+
+        then:
+        runAndAssert("myTask", eagerResult)
+
+        when:
+        buildKotlinFile.text = buildFileDefinition
+            .replace("{inputDeclaration}", "abstract val input: $lazyInputType")
+            .replace("{inputValue}", inputValue)
+
+        then:
+        runAndAssert("myTask", lazyResult)
+
+        where:
+        description                                     | eagerInputType | lazyInputType       | inputValue                               | eagerResult                           | lazyResult
+        "T = T"                                         | "String"       | "Property<String>"  | '"hello"'                                | "Result: hello"                       | "Result: hello"
+        "T = Provider<T>"                               | "String"       | "Property<String>"  | 'provider { \"hello\" }'                 | failsWithDescription("Type mismatch") | "Result: hello"
```

Comment:
🤔 Hm, does Kotlin also support auto-conversion from `Any` to `String`?

:x: In any case, here we also should use a property of type `MyObject`.